### DISCUSSION TITLE
using-quartz-net-with-asp-net-core-and-worker-services/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#50] [Announcements]
# using-quartz-net-with-asp-net-core-and-worker-services/

In this post I show how to run Quartz.NET jobs in ASP.NET Core and worker service apps using the Quartz.Extensions.Hosting package

https://andrewlock.net/using-quartz-net-with-asp-net-core-and-worker-services/

**Comment by steve-sd4d** [comment]
Great article. I've implemented a solution from this that pulls jobs from a database, creates and runs them. How would I go about pausing/restarting a job? All Jobs? I've found some examples online but none seem to tie to this method of creating the jobs using Quartz.

**Comment by sal-420** [comment]
Thanks for posting works great, I could noy have gotten this far without it.

There was one required change to prevent multiple concurrent instance 
Change the task to async and await calling methods:

public TASK Execute(IJobExecutionContext context)
    {
        AWAIT DoSomething();
        return Task.CompletedTask;
    }

**Comment by daruhachker** [comment]
Ok. Nice. But what about clusters? I get an exception if I try to run this code in the Quartz Cluster

**Reply by andrewlock** [reply]
🤷‍♂️ It should work with clusters IIRC!

**Comment by sal-420** [comment]
'var cronSchedule = config[configKey];' returns null  after the scheduler was moved from appsettings.json to appsettings.Development.json and appsettings.Production.json

Any idea as to why 

Thanks

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 02 2020, 04:15, in reply to [Droppi](https://disqus.com/by/droppi/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/50#discussioncomment-6059651)</em>

---

<p>It's the right question to ask :)</p><p></p><blockquote>Won't they bog down web requests, mb better take jobs out to separate app?/blockquote&gt;<br><br>You generally shouldn't see much of an impact on this front. ASP.NET Core is designed to run these hosted services. <br><br>As you mentioned, the benefit of creating a separate app is when you need to scale the services separately. Also, be aware that the design in this post is NON clustered, so if you run multiple instances of the app, you may get contention between jobs running in multiple services at once. I'll write a follow up post on clustering at some point soon :)<br><br>In conclusion, I'd probably keep quartz inside your web app for deployment simplicity until you either need to scale to multiple instances, or want to separate the workloads</blockquote>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 03 2020, 11:19, in reply to [garwin4j](https://disqus.com/by/garwin4j/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/50#discussioncomment-6059657)</em>

---

<p>That's new in the Quartz.Extensions.Hosting package - should definitely help :)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 03 2020, 11:19, in reply to [Oleg Yarem](https://disqus.com/by/olegyarem/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/50#discussioncomment-6059662)</em>

---

<p>That's quite an old document that isn't related to ASP.NET Core :) HostedServices are designed to shut down gracefully, so shouldnt be a problem, even when IIS restarts AFAIK.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 08 2020, 04:06, in reply to [Brad Patton](https://disqus.com/by/disqus_WuXMVdZhLR/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/50#discussioncomment-6059671)</em>

---

<p>I've never used Hangfire, so can't really comment 🙂 🤷‍♂️</p>

[a comment by an automated account is not shown]

**Reply by SuganyadeviKarthik** [reply]
Hi Elias,
Could you please tell me what is the difference between quartz.threadpool.maxconcurrency and quartz.threadpool.threadpoolcount?

Thanks

**Comment by SuganyadeviKarthik** [comment]
Hi,
I want to know what is the difference between quartz.threadpool.maxconcurrency and quartz.threadpool.threadpoolcount?

Can anyone help me.this?

**Comment by guytechNet** [comment]
Hi @andrewlock,

Can I use a jobkey (link to the HelloWorldJob) with multiple WithCronSchedule (Cron Expressions) ?

Example: 

// Create a trigger for the job
                q.AddTrigger(opts => opts
                    .ForJob(jobKey) // link to the HelloWorldJob
                    .WithIdentity("HelloWorldJob-trigger") // give the trigger a unique name
                    .WithCronSchedule("0/5 * * * * ?")); // run every 5 seconds

// Create a trigger for the job
                q.AddTrigger(opts => opts
                    .ForJob(jobKey) // link to the HelloWorldJob
                    .WithIdentity("HelloWorldJob-trigger") // give the trigger a unique name
                    .WithCronSchedule("0/15 * * * * ?")); // run every 15 seconds

// Create a trigger for the job
                q.AddTrigger(opts => opts
                    .ForJob(jobKey) // link to the HelloWorldJob
                    .WithIdentity("HelloWorldJob-trigger") // give the trigger a unique name
                    .WithCronSchedule("0/35 * * * * ?")); // run every 35 seconds
